### DISCUSSION TITLE
How to Properly Integrate Effectful with Hedgehog Using H.evalIO and H.evalM?

### DISCUSSION THREAD

**Discussion by newhoggy** [haskell-effectful/effectful#273] [Q&A] [answered]
I'm working on implementing a custom interpreter for an `effectful` effect that integrates with Hedgehog. The goal is to handle an EvalM operation that allows running computations within the current `Eff` stack and correctly maps them to Hedgehog's evaluation framework using `H.evalM` or `H.evalIO`. Here's my effect definition:

https://github.com/haskell-works/effectful-zoo/blob/main/components/hedgehog/Effectful/Zoo/Hedgehog/Effect/Hedgehog.hs#L21

The interpreter I'm working on looks like this:

https://github.com/haskell-works/effectful-zoo/blob/e2d2087c2c69a11d9b1d3d5d0f4e8dd6f9be9bc1/components/hedgehog/Effectful/Zoo/Hedgehog/Run.hs#L56

I'm running into difficulty because `evalM` takes a callback that is not `IO`:

https://hackage.haskell.org/package/hedgehog-1.5/docs/src/Hedgehog.Internal.Property.html#evalM

What should I do here?

I'm okay if it calls `evalIO` instead, but `evalM` is preferrable.

**Comment by newhoggy** [comment]
I have a work around:

```
    EvalM f -> localUnlift env SeqUnlift $ \unlift -> do
      !result <- try @SomeException $ unlift f

      case result of
        Right a -> pure a
        Left e -> liftTestIO $ H.evalIO $ CE.throw e
```

But I'm still interested to know if the thing I was trying to do is more straightforwardly possible.

**Reply by arybczak** [reply]
Do you have a `MonadTest (Eff es)` instance? If you do, it looks like you should be able to call `evalM` within the unlift.

**Reply by newhoggy** [reply]
Unfortunately, `MonadTest` looks like this:

```
class Monad m => MonadTest m where
  liftTest :: Test a -> m a
```

The argument `TestT Identity` and not `TestT IO`.

**Reply by arybczak** [reply, marked as answer]
Yeah, but you already wrote `liftTest` here: https://github.com/haskell-works/effectful-zoo/blob/main/components/hedgehog/Effectful/Zoo/Hedgehog/Api/Lift.hs, so you have all the components.

It looks to me that you should make `LiftTest` an operation of your effect, then define an instance like described here: https://hackage.haskell.org/package/effectful-core-2.5.0.0/docs/Effectful-Dispatch-Dynamic.html#g:4 and then you can just use `evalM` from the `hedgehog` library with `Eff`.

**Reply by newhoggy** [reply]
Okay, I'm impressed.  Thank you very much!

**Reply by arybczak** [reply]
Also, looks like you're unnecessarily reimplementing tons of functions from `hedgehog`. For example, all functions from https://github.com/haskell-works/effectful-zoo/blob/main/components/hedgehog/Effectful/Zoo/Hedgehog/Api/Assert.hs. All of them in `hedgehog` just need `MonadTest` and you have `MonadTest`, you can just call the originals.

Your `Hedgehog` effect looks to be at a wrong level of granularity. As I suggested, start with it having only `LiftTest`, with this you get `MonadTest (Eff es)` and get the ability to use most of the functions from `hedgehog` with `Eff` for free, there's no need to reimplement them.

You currently hardcoded the `liftTest` function in the `MonadTest (Eff es)` instance, which is the wrong approach. Please follow the guideline in the haddock link I pointed to.

**Reply by newhoggy** [reply]
The `Hedgehog` effect is much smaller now, containing only `LiftTest`, `CatchAssertion` and `ThrowAssertion`.

I had to introduce a new `MonadTestProxy` type class to plumb it up.
